content/technotes/2020-04-30-code-highlighting/index.Rmd

```diff
@@ -0,0 +1,448 @@
+            themeStylesheet.href = '/css/vim.css';
+            themeToggle.innerText = 'Switch to light mode';
+        } else {
+            // if it's dark -> go light
+            themeStylesheet.href = '/css/friendly.css';
+            themeToggle.innerText = 'Switch to dark mode';
+
+        }
+    })
+})
+
+</script>
+<!--/html_preserve-->
+
+
+To generate a stylesheet for a given style, [use Hugo `hugo gen chromastyles --style=monokai > syntax.css` command](https://gohugo.io/content-management/syntax-highlighting/#generate-syntax-highlighter-css). 
+You can then use the stylesheet as is, or tweak it.
+
+How does Chroma know what parts of code is of the string class for instance?
+Once again, regular expressions help, in this case in what is called a lexer. 
+Chroma is inspired by Pygments, and in [Pygments docs](https://pygments.org/docs/quickstart/) it is explained that _"A lexer splits the source into tokens, fragments of the source that have a token type that determines what the text represents semantically (e.g., keyword, string, or comment)."_
+In [R lexer](https://github.com/alecthomas/chroma/blob/master/lexers/r/r.go), ported from Pygments to Chroma by Chroma maintainer [Alec Thomas](https://github.com/alecthomas), for strings we e.g. see
+
+```go
+			{`\'`, LiteralString, Push("string_squote")},
+			{`\"`, LiteralString, Push("string_dquote")},
+// ... code
+		"string_squote": {
+			{`([^\'\\]|\\.)*\'`, LiteralString, Pop(1)},
+		},
+		"string_dquote": {
+			{`([^"\\]|\\.)*"`, LiteralString, Pop(1)},
+		},
+```
+
+Chroma works on Markdown content, so if you use [blogdown to generate pages as html](https://bookdown.org/yihui/blogdown/output-format.html), you can only use client-side highlighting, like [this tidyverse.org page](https://www.tidyverse.org/blog/2017/09/erratum-tidyr-0.7.0/) whose [source is html](https://github.com/tidyverse/tidyverse.org/blob/7ceea9b9bd40207d25e75681041568d600c9d903/content/blog/2017-09-erratum-tidyr/index.html).
+By default nowadays Hugo does server-side syntax highlighting but you could choose to turn it off via [`codeFences = false`](https://gohugo.io/getting-started/configuration-markup#highlight).
+
+We have now seen how Hugo websites have [syntax highlighting, which for Yihui Xie _"is only for cosmetic purposes"_](https://yihui.org/en/2017/07/on-syntax-highlighting/).
+Well, Chroma actually also offers one thing more: line numbering and line highlighting!
+
+### Emphasize parts of your code
+
+With Chroma, you can [apply special options to code blocks defined with fences, i.e. starting with three backticks and language info, and ending with three backticks](https://gohugo.io/content-management/syntax-highlighting/)[^fences].
+
+#### On Chroma options for line highlighting
+
+See how
+
+````
+```r {hl_lines=[1,"4-5"]}
+library("dplyr")
+df %>%
+  mutate(date = lubridate::ymd(date_string)) %>%
+  select(- date_string)
+str(df)
+nrow(df)
+```
+````
```

Comment:
This code block has no language, but the one two down has markdown... That means that in darkmode this block stays white but the other one has a dark mode... are they supposed to be different?

Reply:
no, that's a mistake, good catch!